File: src/verifier.py
import json

class PreconditionError(Exception):
    pass

class InvariantViolationError(Exception):
    pass

def evaluateCondition(condition: str, state: dict):
    # Convert DSL condition to Python syntax
    py_cond = condition.replace(' AND ', ' and ').replace(' OR ', ' or ').replace(' NOT ', ' not ')
    try:
        return eval(py_cond, {}, state)
    except Exception as e:
        raise Exception(f"Error evaluating condition '{condition}': {str(e)}")

def applyMutation(mutation: str, state: dict):
    if '=' not in mutation:
        return
    parts = mutation.split('=')
    v = parts[0].strip()
    expr = '='.join(parts[1:]).strip()
    
    try:
        state[v] = eval(expr, {}, state)
    except Exception:
        # Fallback for simple string assignment
        state[v] = expr.strip('"').strip("'")

def verifyAction(dsl: dict, current_state: dict, action_name: str):
    if action_name not in dsl['actions']:
        raise Exception(f"Action {action_name} not found")
    
    action = dsl['actions'][action_name]

    # 1. Evaluate requires
    for req in action['requires']:
        if not evaluateCondition(req, current_state):
            raise PreconditionError(f"Precondition failed: {req}")

    # 2. Apply mutates
    next_state = current_state.copy()
    for mut in action['mutates']:
        applyMutation(mut, next_state)

    # 3. Evaluate invariants
    for inv in dsl['invariants']:
        if inv['type'] == 'NEVER':
            if evaluateCondition(inv['condition'], next_state):
                error_trace = {
                    "error": "InvariantViolationError",
                    "invariant": f"NEVER {inv['condition']}",
                    "state": next_state
                }
                raise InvariantViolationError(json.dumps(error_trace, indent=2))
        elif inv['type'] == 'ALWAYS':
            if not evaluateCondition(inv['condition'], next_state):
                error_trace = {
                    "error": "InvariantViolationError",
                    "invariant": f"ALWAYS {inv['condition']}",
                    "state": next_state
                }
                raise InvariantViolationError(json.dumps(error_trace, indent=2))

    return next_state

File: src/verifier.ts
export class PreconditionError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'PreconditionError';
  }
}

export class InvariantViolationError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'InvariantViolationError';
  }
}

export function evaluateCondition(condition: string, state: Record<string, any>) {
  const pyCond = condition
    .replace(/\bAND\b/g, '&&')
    .replace(/\bOR\b/g, '||')
    .replace(/\bNOT\b/g, '!')
    .replace(/==/g, '===');
  
  const keys = Object.keys(state);
  const values = Object.values(state);
  try {
    const func = new Function(...keys, `return ${pyCond};`);
    return func(...values);
  } catch (e: any) {
    throw new Error(`Error evaluating condition '${condition}': ${e.message}`);
  }
}

export function applyMutation(mutation: string, state: Record<string, any>) {
  if (!mutation.includes('=')) return;
  const parts = mutation.split('=');
  const v = parts[0].trim();
  const expr = parts.slice(1).join('=').trim();
  
  const keys = Object.keys(state);
  const values = Object.values(state);
  try {
    const func = new Function(...keys, `return ${expr};`);
    state[v] = func(...values);
  } catch (e: any) {
    state[v] = expr.replace(/^"|"$/g, '');
  }
}

export function verifyAction(dsl: any, currentState: Record<string, any>, actionName: string) {
  if (!dsl.actions[actionName]) {
    throw new Error(`Action ${actionName} not found`);
  }
  const action = dsl.actions[actionName];

  for (const req of action.requires) {
    if (!evaluateCondition(req, currentState)) {
      throw new PreconditionError(`Precondition failed: ${req}`);
    }
  }

  const nextState = { ...currentState };
  for (const mut of action.mutates) {
    applyMutation(mut, nextState);
  }

  for (const inv of dsl.invariants) {
    if (inv.type === 'NEVER') {
      if (evaluateCondition(inv.condition, nextState)) {
        const errorTrace = {
          error: "InvariantViolationError",
          invariant: `NEVER ${inv.condition}`,
          state: nextState
        };
        throw new InvariantViolationError(JSON.stringify(errorTrace, null, 2));
      }
    } else if (inv.type === 'ALWAYS') {
      if (!evaluateCondition(inv.condition, nextState)) {
        const errorTrace = {
          error: "InvariantViolationError",
          invariant: `ALWAYS ${inv.condition}`,
          state: nextState
        };
        throw new InvariantViolationError(JSON.stringify(errorTrace, null, 2));
      }
    }
  }

  return nextState;
}
